### PR TITLE
Changes to support EP reshardable per-expert LoRA

### PR DESCRIPTION
**Pull request by jstjohn**
# What does this PR do ?

* Fix EP/DP issues (eg weights getting out of sync between DP ranks when EP<DP) in GroupedExpertLinearAdapter applied to EP layers. 
* Fix EP>1 weight drift in ParallelLinearAdapter applied to EP layers.
* Move peft/utils.py towards new megatron convention of using pg_collections rather than direct calls to parallel_state.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by jstjohn** [comment]
Note, this change appears to work with the case when EP=DP, with/without TP. I am _guessing_ it also works when EP=1 under any DP (have not tested this), but when 1<EP<DP, I have seen evidence that the state still drifts. The pattern I see is that whenever validation happens, the training loss spikes in this case. This is likely due to the weight synchronization that happens right before validation discarding weights that had drifted (maybe between DP ranks?) in that case.

**Comment by jstjohn** [comment]
as of the last commit, this is now working with EP=8, TP=4 on a 64 GPU cluster using `GroupedExpertLinearAdapter`.

File: src/megatron/bridge/peft/utils.py [hunk 1]
```diff
@@ -885,6 +888,207 @@ def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
 
         return x
 
+    def local_experts_per_rank(self) -> int:
+        """Return the number of global expert slots owned by this EP rank."""
+
+        ep_size = parallel_state.get_expert_model_parallel_world_size() or self.config.expert_model_parallel_size or 1
```
Comment (yaoyu-33) [inline src/megatron/bridge/peft/utils.py]: can we avoid using parallel_state, we can get the information from PG collection attribute, we can pass in as kwargs. MCore is refactor to avoid parallel state in general to make it easier for encoder / decoder have different mp groups. we can pass in required pgs.

File: src/megatron/bridge/peft/utils.py [hunk 2]
```diff
@@ -906,40 +1110,34 @@ def sharded_state_dict(
             Sharded state dictionary for distributed checkpointing.
         """
         sharded_state_dict = {}
+        use_expert_axis = self._uses_grouped_expert_sharding()
```
Comment (yaoyu-33) [inline src/megatron/bridge/peft/utils.py:1250]: can we add comments here for what is and when to use "use_expert_axis" for future clarity.

File: src/megatron/bridge/peft/utils.py (hunk 1 above)
Comment (yaoyu-33) [reply src/megatron/bridge/peft/utils.py]: We can get pg_collection from parallel state for now, but for those utils func, let's just pass in the comm groups instead of using paralell state constantly.

Comment (jstjohn) [reply src/megatron/bridge/peft/utils.py]: I actually just finished a more complete rewrite for peft to use pg_collection. What do you think?

File: src/megatron/bridge/peft/utils.py (hunk 2 above)
Comment (jstjohn) [reply src/megatron/bridge/peft/utils.py:1250]: done

**Review by yaoyu-33 — DISMISSED** [review]

**Comment by JRD971000** [comment]
@jstjohn CI is failing, could youn please TAL?

**Comment by jstjohn** [comment]
Taking a look

**Comment by jstjohn** [comment]
```
 tests/unit_tests/data/datasets/test_chat_template.py::TestPackedSequenceWithChatEndToEnd::test_tokenize_dataset_produces_loss_mask 
    0% 0/1 [00:00<?, ?it/s]
  100% 1/1 [00:00<00:00, 1119.97it/s]
  Stack (most recent call first):
    File "<frozen posixpath>", line 476 in _joinrealpath
    File "<frozen posixpath>", line 432 in realpath
    File "/opt/venv/lib/python3.12/site-packages/coverage/files.py", line 158 in abs_file
    File "/opt/venv/lib/python3.12/site-packages/coverage/files.py", line 85 in canonical_filename
    File "/opt/venv/lib/python3.12/site-packages/coverage/inorout.py", line 401 in should_trace
    File "/opt/venv/lib/python3.12/site-packages/coverage/control.py", line 454 in _should_trace
    File "/usr/lib/python3.12/concurrent/futures/process.py", line 94 in _python_exit
    File "/usr/lib/python3.12/threading.py", line 1592 in _shutdown
    File "/usr/lib/python3.12/multiprocessing/process.py", line 332 in _bootstrap
    File "/opt/venv/lib/python3.12/site-packages/coverage/multiproc.py", line 51 in _bootstrap
    File "/usr/lib/python3.12/multiprocessing/popen_fork.py", line 71 in _launch
  Stack (most recent call first):
  Stack (most recent call first):
    File   File ""Stack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.pyStack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py  File "  File Stack (most recent call first):
  Stack (most recent call first):
  "  File Stack (most recent call first):
  "  File , line "  File , line "Stack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py""58, line /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py"221/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py  File /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py in 58"/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py in "", line "_connect in , line   File "executescript/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py122, line 
  _connect221" in , line 
  " in 221  File 
  /usr/lib/python3.12/multiprocessing/popen_fork.pyexecutescript221  File , line _execute in "  File "
   in "122
  executescript/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py", line Stack (most recent call first):
    File   File executescript/opt/venv/lib/python3.12/site-packages/coverage/sqldata.py in   File Stack (most recent call first):
  
  "/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py19"Stack (most recent call first):
  "
  "_execute"Stack (most recent call first):
    File   File , line "Stack (most recent call first):
   in /opt/venv/lib/python3.12/site-packages/coverage/sqldata.pyStack (most recent call first):
    File /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py  File , line 
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py  File ""96, line   File __init__"  File Stack (most recent call first):
  """353  File ""/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py/opt/venv/lib/python3.12/site-packages/coverage/sqldata.py in 96"
  , line   File 353""  File /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py, line /opt/venv/lib/python3.12/site-packages/coverage/sqldata.py in ", line /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py""__enter__ in /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py in /usr/lib/python3.12/multiprocessing/context.py_init_db/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py""221Stack (most recent call first):
  "  File _init_dbStack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py158", line Stack (most recent call first):
  , line   File 
  __enter__"Stack (most recent call first):
  "
  "/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py, line  in , line "
    File " in Stack (most recent call first):
  , line 58353"  File Stack (most recent call first):
  
  , line   File , line Stack (most recent call first):
    File   File , line "58executescript353Stack (most recent call first):
  Stack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py  File ", line execute_void  File 58Stack (most recent call first):
   in  in Stack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/sqldata.py"Stack (most recent call first):
  Stack (most recent call first):
    File   File "  File Stack (most recent call first):
  58Stack (most recent call first):
  "Stack (most recent call first):
  Stack (most recent call first):
  282" in "Stack (most recent call first):
  58Stack (most recent call first):
  , line Stack (most recent call first):
  Stack (most recent call first):
  Stack (most recent call first):
  Stack (most recent call first):
   in   File 
   in   File   File ""Stack (most recent call first):
  Stack (most recent call first):
  Stack (most recent call first):
  "", line /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py158Stack (most recent call first):
  
  " in Stack (most recent call first):
  Stack (most recent call first):
    File _connect_init_db  File 
  ""/opt/venv/lib/python3.12/site-packages/coverage/sqldata.pyStack (most recent call first):
  Stack (most recent call first):
    File Stack (most recent call first):
  Stack (most recent call first):
  "Stack (most recent call first):
  Stack (most recent call first):
  Stack (most recent call first):
  <frozen posixpath>Stack (most recent call first):
  Stack (most recent call first):
  "Stack (most recent call first):
  Stack (most recent call first):
    File  in Stack (most recent call first):
  Stack (most recent call first):
    File Stack (most recent call first):
  Stack (most recent call first):
  <frozen posixpath>  File "  File   File /opt/venv/lib/python3.12/site-packages/coverage/sqldata.py_Popen"Stack (most recent call first):
  Stack (most recent call first):
  /opt/venv/lib/python3.12/site-packages/coverage/misc.py  File  in Stack (most recent call first):
  Stack (most recent call first):
    File   File "Stack (most recent call first):
  58Stack (most recent call first):
    File Stack (most recent call first):
    File   File _connect""  File Stack (most recent call first):
  _init_db/opt/venv/lib/python3.12/site-packages/coverage/sqldata.py<frozen posixpath>"Stack (most recent call first):
    File Stack (most recent call first):
    File Stack (most recent call first):
    File /opt/venv/lib/python3.12/site-packages/coverage/sqldata.py58" in , line Stack (most recent call first):
    File   File /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py_connect  File 
  "  File "Stack (most recent call first):
  Stack (most recent call first):
  
  Stack (most recent call first):
    File <frozen posixpath>, line "Stack (most recent call first):
    File   File Stack (most recent call first):
  Stack (most recent call first):
  "Stack (most recent call first):
  Stack (most recent call first):
    File   File <frozen posixpath>  File   File ""  File Stack (most recent call first):
    File   File /opt/venv/lib/python3.12/site-packages/coverage/sqldata.py  File   File "_connect  File   File "Stack (most recent call first):
    File   File ", line /opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.py476""Stack (most recent call first):
  
  , line   File 328  File   File ""Stack (most recent call first):
  _connect  File Stack (most recent call first):
  "<frozen posixpath>/opt/venv/lib/python3.12/site-packages/coverage/sqlitedb.pyStack (most recent call first):
    File  in Stack (most recent call first):
    File Stack (most recent call first):
  ""  File "Stack (most recent call first):
  ```
  It seems related to coverage collection, and maybe the unit tests in `tests/unit_tests/data/datasets/test_chat_template.py::TestPackedSequenceWithChatEndToEnd::test_tokenize_dataset_produces_loss_mask`.

**Comment by jstjohn** [comment]
@JRD971000 FYI fixed the issue that was blocking the initial unit test phase of CI, 🤞 on the full stage.

**Comment by jstjohn** [comment]
@JRD971000 looks like it's all passing now.

**Review by yaoyu-33 — APPROVED** [review]
